### ISSUE TITLE
Question: The best practices to update the model of collection cell

### ISSUE THREAD

**Issue by smalllixin** [facebook/componentkit#114]
Here's the logic to maintain the model's consistency I thought:
1. Register/unregister an observer for the model by ID.(in ComponentController lifecycle methods)
2. If some model modified, will get a notification and a new model copy.

However I found the only way to update the model is use CKArrayControllerInputItems(update). 
So I have to make a map<ID, indexpath> and tracking all model's update. 
What is the best practise to update the model in componentkit.

Thanks in advance.

**Comment by smalllixin** [comment]
![q](https://cloud.githubusercontent.com/assets/787881/7151503/a7f25c28-e35d-11e4-954d-e7a33808f5e3.jpg)

Pardon me. Maybe my expression is not good. 
The image show the issue I meet.

**Comment by dstnbrkr** [comment]
Is the double refresh the issue?

**Comment by smalllixin** [comment]
@dstnbrkr yes

**Comment by dstnbrkr** [comment]
It looks like the tap action triggers both a model update and a state change / animation. Instead, could that tap action trigger only a model update (no state change)? Then the animation could occur in
`animationsFromPreviousComponent` when the updated component arrives.

Since that model update is async, the resulting animation might not happen soon enough. To update the view sooner than the model update refresh, take a look at CKOptimisticViewMutations.h.

**Comment by smalllixin** [comment]
Got it. Thank u.

**Comment by smalllixin** [comment]
@dstnbrkr 
Now I try tap action only perform mode update. However, I found animationsFromPreviousComponent is only called after updateState, the new model update do not trigger that.

**Comment by dstnbrkr** [comment]
@smalllixin I'd expect a model update to cause the component to remount and animate. If you break in componentWillMount [1], does it take a path that does not call animationsFromPreviousComponent?

[1] https://github.com/facebook/componentkit/blob/568d14f71b6d4d92bf14ea10579be587223fc7a2/ComponentKit/Core/CKComponentController.mm#L94

**Comment by smalllixin** [comment]
@dstnbrkr 
After made some tests. If use hostingView , assign a new model , then animationsFromPreviousComponent called correctly.
```Objective-C
  switch (_state) {
    case CKComponentControllerStateUnmounted:
      _state = CKComponentControllerStateMounting;
      [self willMount];
      break;
    case CKComponentControllerStateMounted://!!! state === Mounted
      _state = CKComponentControllerStateRemounting;
      [self willRemount];
      if (_previousComponent) { // Only animate if updating from an old component to a new one, and previously mounted
        for (const auto &animation : [component animationsFromPreviousComponent:_previousComponent]) {
          _pendingAnimations.push_back({animation, animation.willRemount()});
        }
      }
      break;
    default:
      CKFailAssert(@"Unexpected state '%@' in %@ (%@)", componentStateName(_state), [self class], _component);
  }
```
However, in collectionview. I update the model using the follow code. 
```Objective-C
CKArrayControllerInputItems items;
items.update([NSIndexPath indexPathForRow:0 inSection:0], newLike);
[_dataSource enqueueChangeset:{{}, items}
                  constrainedSize:[_sizeRangeProvider sizeRangeForBoundingSize:self.collectionView.bounds.size]];
```
The different between hosting state become unmounted:
```Objective-C
  switch (_state) {
    case CKComponentControllerStateUnmounted://!!!! state == umounted this time
      _state = CKComponentControllerStateMounting;
      [self willMount];
      break;
    case CKComponentControllerStateMounted:
      _state = CKComponentControllerStateRemounting;
      [self willRemount];
      if (_previousComponent) { // Only animate if updating from an old component to a new one, and previously mounted
        for (const auto &animation : [component animationsFromPreviousComponent:_previousComponent]) {
          _pendingAnimations.push_back({animation, animation.willRemount()});
        }
      }
      break;
    default:
      CKFailAssert(@"Unexpected state '%@' in %@ (%@)", componentStateName(_state), [self class], _component);
  }
```

**Comment by dstnbrkr** [comment]
@benlodotcom know what's going on with this difference in mount/remount behavior between hosting view and collection view cell?

**Comment by itsthejb** [comment]
Hmmm... Any resolution to this? I appear to be experiencing the same issue. I have a model change that creates a new model copy (also a Like state change). I'm triggering an update with `CKArrayControllerInputItems`. The new component is correctly created, but I don't get the right lifecycle  method sequence to trigger an animation. I have a parent and child component where the child is the like button. They are both scoped with the same model object identifier, although they are different `CKCompositeComponent` subclasses.

Sequence:

1. Trigger update in data source with `CKArrayControllerInputItems`
2. `new...` method for parent component
3. `new...` method for child component
4. `willUnmount...` for child controller (`CKComponentControllerStateMounted`)
5. `didUnmount...` for child (`CKComponentControllerStateUnmounting`)
6. `willUnmount...` for parent (`CKComponentControllerStateMounted`)
7. `didUnmount...` for parent (`CKComponentControllerStateUnmounting`)
8. `willMount...` for parent, new component (`_updatingComponent = YES`)
9. `willMount...` for child, new component (`_updatingComponent = YES`)
10. `didMount...` for child (`CKComponentControllerStateMounting
...

So basically I don't get component animation calls because both components are `CKComponentControllerStateUnmounted` when `componentWillMount:` is called (steps 4-7). Any ideas why I'm getting this? Thanks!

**Comment by dstnbrkr** [comment]
@adamjernst looks like animations require a remount, but the conditions that trigger a remount vs. a mount are tough to get right (and might be different depending on the container). Got any ideas on how to debug the case we get a mount when we expect a remount (or an alternate trigger for animations)?

**Comment by crylico** [comment]
What if the check to append animations were outside the switch statement?  This worked without a problem in my situation, since there's already an if condition to make sure there's a previous component.  

Unsure though if this could cause a problem

**Comment by itsthejb** [comment]
Any chance on some feedback as to whether or not this is an error on my part or an edge case in the library? Since this is pretty core functionality I'm quite inclined to believe that I'm doing something wrong. However, updating CK with a model property change is quite a fundamental function which isn't really directly addressed in the documentation.

Also I have a deadline swiftly approaching... :sweat_smile:

**Comment by dstnbrkr** [comment]
I wouldn't rule out bug, we're at the mercy of collection view or table view here. There's a chance the container could reload the row underneath us.

So we need a remount to trigger the animation, but we're getting a mount. We only get a mount when the component was unmounted, so the next step is to find out why it's unmounted and then debug from there.

@itsthejb can you break in unmount and share the backtrace?

**Comment by itsthejb** [comment]
No probs. Parent component:

```
* thread #1: tid = 0x1dba850, 0x000000010c33a21a <REDACTED>`-[CKComponentController componentWillUnmount:](self=0x00007faa4cfc60e0, _cmd=0x000000010c8723c5, component=0x00007faa4e2555d0) + 42 at CKComponentController.mm:134, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
  * frame #0: 0x000000010c33a21a <REDACTED>`-[CKComponentController componentWillUnmount:](self=0x00007faa4cfc60e0, _cmd=0x000000010c8723c5, component=0x00007faa4e2555d0) + 42 at CKComponentController.mm:134
    frame #1: 0x000000010c324d44 <REDACTED>`-[CKComponent unmount](self=0x00007faa4e2555d0, _cmd=0x000000010c872386) + 180 at CKComponent.mm:155
    frame #2: 0x000000010c3640a0 <REDACTED>`-[CKComponentLifecycleManager detachFromView](self=0x00007faa4e23fdb0, _cmd=0x000000010c872e7f) + 848 at CKComponentLifecycleManager.mm:181
    frame #3: 0x000000010c363c49 <REDACTED>`-[CKComponentLifecycleManager attachToView:](self=0x00007faa4e23fdb0, _cmd=0x000000010c872153, view=0x00007faa4e079ff0) + 137 at CKComponentLifecycleManager.mm:168
    frame #4: 0x000000010c32145a <REDACTED>`-[CKCollectionViewDataSource collectionView:cellForItemAtIndexPath:](self=0x00007faa4cef7a10, _cmd=0x000000010fe6d298, collectionView=0x00007faa4b0b2c00, indexPath=0xc000000000008016) + 554 at CKCollectionViewDataSource.mm:129
    frame #5: 0x000000010fcac0c7 UIKit`-[UICollectionView _createPreparedCellForItemAtIndexPath:withLayoutAttributes:applyAttributes:] + 235
    frame #6: 0x000000010fcc0f67 UIKit`__51-[UICollectionView _viewAnimationsForCurrentUpdate]_block_invoke1381 + 508
    frame #7: 0x000000010fcbedce UIKit`-[UICollectionView _viewAnimationsForCurrentUpdate] + 3956
    frame #8: 0x000000010fcc1e24 UIKit`-[UICollectionView _updateWithItems:tentativelyForReordering:] + 2075
    frame #9: 0x000000010fcbd9b4 UIKit`-[UICollectionView _endItemAnimationsWithInvalidationContext:tentativelyForReordering:] + 13016
    frame #10: 0x000000010fcc37ae UIKit`-[UICollectionView _performBatchUpdates:completion:invalidationContext:tentativelyForReordering:] + 437
    frame #11: 0x000000010c3219ef <REDACTED>`-[CKCollectionViewDataSource componentDataSource:hasChangesOfTypes:changesetApplicator:](self=0x00007faa4cef7a10, _cmd=0x000000010c87217b, componentDataSource=0x00007faa4cef15f0, changeTypes=64, changesetApplicator=0x00007fff53f3fd58) + 239 at CKCollectionViewDataSource.mm:173
    frame #12: 0x000000010c3429a2 <REDACTED>`-[CKComponentDataSource _processChangeset:](self=0x00007faa4cef15f0, _cmd=0x000000010c872a86, changeset=0x00007fff53f40060) + 690 at CKComponentDataSource.mm:376
    frame #13: 0x000000010c342631 <REDACTED>`-[CKComponentDataSource _componentPreparationQueueDidPrepareBatch:sections:](self=0x00007faa4cef15f0, _cmd=0x000000010c872a16, batch=0x00007faa49e2fda0, sections=0x00007faa4e239ac0) + 1745 at CKComponentDataSource.mm:324
    frame #14: 0x000000010c341d5f <REDACTED>`__43-[CKComponentDataSource _enqueueChangeset:]_block_invoke115(.block_descriptor=0x00007faa4e275f80, sections=0x00007faa4e239ac0, ID=1, outputBatch=0x00007faa49e2fda0, isContiguousTailInsertion=false) + 607 at CKComponentDataSource.mm:272
    frame #15: 0x000000010c36b2c9 <REDACTED>`__43-[CKComponentPreparationQueue _processJob:]_block_invoke165(.block_descriptor=0x00007faa49e7b0a0) + 265 at CKComponentPreparationQueue.mm:229
    frame #16: 0x0000000112b36186 libdispatch.dylib`_dispatch_call_block_and_release + 12
    frame #17: 0x0000000112b55614 libdispatch.dylib`_dispatch_client_callout + 8
    frame #18: 0x0000000112b3da1c libdispatch.dylib`_dispatch_main_queue_callback_4CF + 1664
    frame #19: 0x00000001120571f9 CoreFoundation`__CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
    frame #20: 0x0000000112018dcb CoreFoundation`__CFRunLoopRun + 2043
    frame #21: 0x0000000112018366 CoreFoundation`CFRunLoopRunSpecific + 470
    frame #22: 0x0000000113af4a3e GraphicsServices`GSEventRunModal + 161
    frame #23: 0x000000010f68d900 UIKit`UIApplicationMain + 1282
    frame #24: 0x000000010bcc02e1 <REDACTED>`main(argc=1, argv=0x00007fff53f413b8) + 81 at main.m:15
    frame #25: 0x0000000112b89145 libdyld.dylib`start + 1
    frame #26: 0x0000000112b89145 libdyld.dylib`start + 1
```

Looks like this is enough?

**Comment by adamjernst** [comment]
Yup the problem is in `CKCollectionViewDataSource`:

https://github.com/facebook/componentkit/blob/master/ComponentKit/DataSources/CKCollectionViewDataSource.mm#L240

Calling `reloadItemsAtIndexPaths:` results in an entirely new cell being used by the `UICollectionView` in most cases. Thus we get unmount + mount instead of remount.

In the data source we use internally (which isn't the open source version for various reasons) we instead call `cellForItemAtIndexPath:` and then just reconfigure the cell directly, instead of requiring the `UICollectionView` to get an entirely new cell. This is more efficient and will cause a remount instead.

So — bug, but one that you could work around by forking the data source for now. I'll try to make sure we fix this in the new collection view data source @benlodotcom  is working on.

**Comment by adamjernst** [comment]
I'm going to close this since it's a question. The fix will be tracked under #193.

**Comment by itsthejb** [comment]
Ok, thanks. I'll look into doing a temporary fix and possibly post what I come up with.
